### PR TITLE
Document hotfix deployment instructions

### PR DESCRIPTION
**Pull request by stevehook**
## Context

This is an incident action after the 31 January incident. We want to have an agreed process for deploying hotfixes.

## Changes proposed in this pull request

- [x] Add a new _hotfix deployment_ document that outlines the process and provides instructions.
- [x] Link to the new document from the normal deployment instructions.
- [x] Update the Azure deployment pipeline to permit deployments from a branch called `hotfix`

Beyond the scope of this PR we also need to update the ops dashboard to include a warning when a hotfix deployment is running. Perhaps this could be a simple banner triggered by the presence of a `hotfix` branch in the Github repo? We may also look at tagging releases in the future.

## Guidance to review

- Are the instructions clear? Could they be followed under pressure?
- Any situations that we've not foreseen?

## Link to Trello card

https://trello.com/c/2c9pgmt8/1622-figure-out-how-to-do-a-proper-hotfix-without-deploying-other-stuff-on-master-%F0%9F%8F%88

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: docs/hotfix-deployment.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+feature1           -x-x-x     /   \          /
+                   \         /     \        /
+feature2            --x-x-x-x       \      /
+                                     \    /
+feature3                              -x-x
+```
+
+### Some questions
+
+Do we need to tag releases? At the moment we don't really have record of
```
Comment (mnacos) [inline docs/hotfix-deployment.md]: +1 for tagging releases

File: docs/hotfix-deployment.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+which versions shipped and when in the git repository itself. It makes
+sense to tag releases that ship to production so that we have a readily
+available record of that. It would make finding the commit from which a
+hotfix should be branched a little easier.
+
+Should we get into the habit of creating a release branch for normal
+releases? (A release branch is really a generalisation of a hotfix
+branch)
+
+How can we automate step 4 (merging the hotfix back to master)? We want
```
Comment (mnacos) [inline docs/hotfix-deployment.md]: Hotfixes are meant to be rare events, so asking the hotfixer to raise a PR for merging hotfix into master seems fine to me. We could even leverage the need to create a PR as a confirmation step that the hotfix has worked. Finally, we should probably extend the apply ops dashboard to detect if a hotfix is in progress.

File: docs/hotfix-deployment.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+to make sure we don't lose any hotfixes in a subsequent deploy.
+
+Are there any branch naming or other conventions that we want to follow?
+It would be nice not to have to change the current convention for naming
+feature branches.
+
+Issues that are not addressed here:
+
+#### What if we had more than one hotfix on the go at a time?
+The main problem here would be multiple hotfix branches. If we can stick
```
Comment (mnacos) [inline docs/hotfix-deployment.md]: +1 for single hotfix branch

File: docs/hotfix-deployment.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+`master`. We urgently need to create a fix that does not include
+features 1, 2 and 3, we just want everything up to `D`.
+
+### Instructions
+
+1. Inform the team that a hotfix is being prepared by posting a new
+   thread in Slack to #twd_apply.
+2. Fetch `master` to your local machine and create a branch called
+   `hotfix` from the last deployed commit (not `HEAD` as we normally
+   would). You can specify the last deployed commit using it's SHA and
```
Comment (duncanjbrown) [inline docs/hotfix-deployment.md]:
```suggestion
   would). You can specify the last deployed commit using its SHA and
```

File: docs/hotfix-deployment.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+  ```
+  $ git fetch origin master
+  $ git checkout -b hotfix <sha-from-ops-dashboard>
+  $ git push origin hotfix
+  ```
+
+3. Implement the fix locally raise a PR and get it approved in the
+   normal way. Test it locally or using the Heroku review app that is
+   be automatically created.
```
Comment (malcolmbaig) [inline docs/hotfix-deployment.md]: typo: 'be'

File: docs/hotfix-deployment.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+   [ops dashboard](https://apply-ops-dashboard.herokuapp.com/) - just
+   look for the current production version. Push the new branch to Github.
+
+  ```
+  $ git fetch origin master
+  $ git checkout -b hotfix <sha-from-ops-dashboard>
+  $ git push origin hotfix
+  ```
+
+3. Implement the fix locally raise a PR and get it approved in the
```
Comment (duncanjbrown) [inline docs/hotfix-deployment.md]:
```suggestion
3. Implement the fix locally, raise a PR and get it approved in the
```

File: docs/manual-deployment.md [hunk 7]
```diff
@@ -10,6 +10,10 @@ In the event that code changes are made to the app but the Azure pipelines fail
 
 This process assumes that the build and test stage has completed without error and a Docker image has been uploaded to Dockerhub successfully. This process also assumes that no changes are expected to be made to the configuration of the underlying Azure infrastructure hosting the application.
 
+Note that these instructions are for normal deployment. For a hotfix
+deployment, for an urgent fix, please refer to [Hotfix deployment
```
Comment (duncanjbrown) [inline docs/manual-deployment.md]:
```suggestion
deployment, please refer to [Hotfix deployment
```

File: docs/hotfix-deployment.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+   First deploy the `hotfix` branch using the [normal deployment
+   procedure](deployment.md) (except picking `HEAD` of the
+   `hotfix` branch rather than a specific SHA on `master` as we normally
+   do).
+5. Merge the `hotfix` branch back to `master`. At this point the `hotfix`
+   branch should be automatically deleted.
+
+Here is an example of the process where the last 'normal' deployment was
+at `D` and the hotfix was deployed at `H`. After the `hotfix` branch is
+deployed back into `master` it's back to business as usual.
```
Comment (malcolmbaig) [inline docs/hotfix-deployment.md]: `merged back into` ?

File: docs/hotfix-deployment.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                  \      /     / \            /
+feature1           -x-x-x     /   \          /
+                   \         /     \        /
+feature2            --x-x-x-x       \      /
+                                     \    /
+feature3                              -x-x
+```
+
+### Notes
+- We always use the name `hotfix` to enforce the rule that only
```
Comment (malcolmbaig) [inline docs/hotfix-deployment.md]: `that we only`

**Comment by malcolmbaig** [comment]
LG, just some minor typos

**Comment by stevehook** [comment]
> Makes sense to me! Love the ASCII diagrams, really helps to visualise what we want to do here. 💯
> 
> Will leave approve to others with more experience.

thanks @chubberlisk - from my point of view your approval (or not) carries as much weight as any other!

**Pull request merged**